### DISCUSSION TITLE
Remove page with on-demand Revalidation

### DISCUSSION THREAD

**Discussion by dev-fredericfox** [vercel/next.js#37093] [Help]
Is it possible to delete/remove a path/page with on-demand Revalidation?

Say I have a this structure : `/pages/[movies].js`
Let's say right now 3 movies are available:
```
/pages/lotr
/pages/harryPotter
/pages/StarWars
```
Is it there a way to remove one movie with res.unstable_revalidate() or some other workaround?

I tried to search the discussions but couldn't find if it is possible to do, and how.

**Comment by Ali-Hussein-dev** [comment]
Is it a dynamic route? for example `pages/[...]`?

**Reply by dev-fredericfox** [reply]
Yes, it is!

**Comment by Ali-Hussein-dev** [comment]
"Is it possible to delete/remove a path/page with on-demand Revalidation?"
@dev-fredericfox when you revalidate, the next server rebuilds the page and sends a new copy to the edge server. 
here is the revalidation implementation, this will be triggered when you want to remove a page from CDN and replace it with a new copy.
```js
// pages/api/revalidate.js

export default async function handler(req, res) {
  // Check for secret to confirm this is a valid request
  if (req.query.secret !== process.env.MY_SECRET_TOKEN) {
    return res.status(401).json({ message: 'Invalid token' })
  }
  try {
    await res.unstable_revalidate('/pages/lotre')
    return res.json({ revalidated: true })
  } catch (err) {
    // If there was an error, Next.js will continue
    // to show the last successfully generated page
    return res.status(500).send('Error revalidating')
  }
}
```

**Comment by stormsson** [comment]
@Ali-Hussein-dev I think there is some info missing

When rebuilding for a page that's now missing it's my impression that we fall in the `catch` branch of the try/catch.
for example I have something like:

`error - Error: Failed to revalidate /removed-page: Invalid response 404`

@dev-fredericfox 
At this point of the process the rebuild already happened though: if you check /removed-page you'll reach a 404.

what could probably be improved is to check the catch error in the handler: if it's a 404 message just don't raise it.

Warning though; this could lead to unraised / unseen exceptions when you expect an existing page to be rebuilt and, for some reason, the router can't find it

**Reply by Ali-Hussein-dev** [reply]
> 
I don't think returning 404 means the rebuild happened, it means you have a typo with the provided path `unstable_revalidate(path with typo)` or the page query does not exist.

**Reply by stormsson** [reply]
Well, but the effects are the one expected: a page that was built and accessible before it's not accessible anymore and returns 404.

I also believe that it is correct that the answer is a 404: in this scenario after depublishing the page from the cms, the specific page path is not returned anymore to getStaticPaths.

Calling the revalidation actually generates the 404 page for that slug

**Reply by advename** [reply]
Can confirm, on-demand revalidation, if handled correctly, `getStaticProps` can remove a deleted page.

With handled correctly is meant that `getStaticPaths` should use `"blocking"` or `true` as `fallback`, and `getStaticProps` should check if a requested slug exist, and if not, return
```js
return {
    notFound: true,
}
```
